llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp

```diff
@@ -1738,6 +1738,7 @@ Instruction *InstCombinerImpl::foldICmpAndShift(ICmpInst &Cmp,
 
     // Compute X & (C2 << Y).
     Value *NewAnd = Builder.CreateAnd(Shift->getOperand(0), NewShift);
+    Cmp.setSameSign(false);
     return replaceOperand(Cmp, 0, NewAnd);
```

Comment:
Can we create a new icmp here instead? I don't see a good reason to do in-place modification here, and it causes this kind of issue...